### PR TITLE
Integrate and Enhance Type Checking with Mypy

### PR DESCRIPTION
**Pull request by gao-hongnan**
This PR introduces several enhancements and changes focusing on integrating and utilizing `mypy` for improved type checking and code quality. Below are the key points of this PR:

✅ **Mypy Configuration File**: Closely aligned with the [OpenAI Python mypy settings](https://github.com/openai/openai-python/blob/main/mypy.ini). 

✅ **Pin Mypy Versions**: The versions of `mypy` are pinned.

✅ **Mypy Pre-commit Implementation**: Integrated `mypy` into the pre-commit hooks.

✅ **Mypy GitHub Actions**: Implemented GitHub Actions for running `mypy` checks. 

✅ **Mypy Artifacts Logs for Persistency**: Configured the workflow to generate and store `mypy` logs as artifacts. This provides a persistent record of the type checks for each build, useful for debugging and historical reference.

✅ **Fixed Mypy Errors in `cli/cli.py` and `cli/usage.py`**: As a proof of concept (POC), type errors in `cli/cli.py` and `cli/usage.py` have been fixed. Other errors can be fixed gradually as a collective effort.

Some screenshots.
Screenshot 2023-12-09 at 11 08 39 PM

Screenshot 2023-12-09 at 11 09 11 PM

As future enhancement, we can group `ruff`, `mypy`, `test` etc under a single workflow in github actions.

## Summary by CodeRabbit

- **New Features**
  - Added a new continuous integration workflow to enhance code quality and reliability.
  - Added a type alias `ModelNames` for improved type checking and clarity in code.
  - Implemented new static type checks using mypy for specific Python files.

- **Documentation**
  - Updated type hints and return types in various functions for better code documentation and type safety.

- **Refactor**
  - Removed the `ruff-format` hook from the pre-commit configuration.
  - Enhanced import statements with additional type hints.

- **Dependencies**
  - Added `mypy` version `1.7.1` to dependencies for static type checking.
  - Included `typer` in the list of dependencies for creating command-line interfaces.
  - Removed `ruff` version `0.1.7` from dependencies.

- **Bug Fixes**
  - Adjusted mypy configuration to enforce stricter type checking and resolve potential type-related issues.

### CODE REVIEW

**Review by jxnl — APPROVED** [review]
Looks good. Are you planning to add more before merge?

File: .github/workflows/mypy.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+name: Continuous Integration
```
Comment (jxnl) [inline .github/workflows/mypy.yml:1]: Use a better name

Comment (gao-hongnan) [reply .github/workflows/mypy.yml:1]: @jxnl Renamed from `Continuous Integration` to `MyPy` for better readability.

**Comment by gao-hongnan** [comment]
> Looks good. Are you planning to add more before merge?

If this PR looks decent to you, I can help set up a more complete suite of pre-merge continuous integration (CI) checks.

**Comment by jxnl** [comment]
would love that!

**Comment by gao-hongnan** [comment]
> would love that!

Got it, I will open a new PR for that. For now, I am finding out why the latest `test/coverage` check is unsuccessful on this branch.

**Comment by gao-hongnan** [comment]
Hi @jxnl, after inspecting the errors and digging around and seeing all failed cases have exactly the same error:

```txt
openai.APIConnectionError: Connection error.
```

which may be related to the secrets `OPENAI_API_KEY` defined in the github actions.

Consequently, it could be because GitHub Actions provides restricted access to secrets when running workflows triggered by pull requests from forks. This is further confirmed after I checked [stackoverflow](https://stackoverflow.com/questions/76952023/how-to-make-github-actions-safely-access-secrets-for-prs-created-from-forks) and a [blog post from GitHub](https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/). The blog post mentioned it can be somewhat (unsure) resolved by `pull_request_target`.

**Comment by jxnl** [comment]
that makes sense running tests now

**Comment by gao-hongnan** [comment]
> that makes sense running tests now

You might have to first define the following in `.github/workflows/test.yml`:

```yaml
pull_request_target:
  branches:
    - main
 ```

But after reading a bunch of articles, there are still potential risks involved, see [here](https://nathandavison.com/blog/github-actions-and-the-threat-of-malicious-pull-requests). As of now I do not have a better alternative given that the recent changes also depends on `${{ secrets.OPENAI_API_KEY }}` when running certain tests.

**Review by jxnl — APPROVED** [review]
+1

**Pull request merged**